### ISSUE TITLE
Backpatch fix to tailcall devirtualization bug for 2.1 LTS

### ISSUE THREAD

**Issue by Tomius** [dotnet/coreclr#22385]
tl;dr: There's a bug in dotnet core ```[2.1, 3.0-preview)``` which is already fixed on master, but it's severe enough that it should be backpatched to the long term support branch.

## Bug description
Unboxing during devirtualization is done via creating a local copy of the struct on the stack, and using the memory address of the local copy as the ```this``` pointer. When tail calls are also involved in this process, the stack frame containing the local copy gets unwinded before the devirtualised function is called, so the ```this``` points to a value on the stack that is no longer there.

## Previous work relating this bug
This bug was fixed on master by #21804, and there's also a closed issue about this:  #21890

### It's a security issue
The bug is triggering an undefined behaviour from a safe context, which could easily become a security issue.

Consider this simple F# "Hello World" example:
``` F#
open System

[<Struct>]
type A (value : string, __ : int64) =
  override this.ToString() : string = value.ToString()

type B (a : A) =
  override this.ToString() : string = (a :> Object).ToString()

[<EntryPoint>]
let main args =
  Console.WriteLine(B(A("hello world", 4L)))
  0
```

Now note one of its possible outputs (it's not deterministic):
```
> dotnet run --fx-version 2.1.3 -c Release
????N ??N ???  ? ??? ??N ??? ??N ??? H   ??N N     ??N        ??N ??N ??N  ?      ? ??? ??N
???      ??? ???    ??N ??? ??? ??N ??? ? ??? ???  ?   ? ?   ??? ?? ??N ??N            ??? ???
??N ??N  ? ???   ? ??? ??? ??N ??N ???         ??? ??N    ???       ?       ? ???    ???
           ???          ??? ??? ??? ???       ?       ? ???    ???     ???    ???      
??N            P ?            ??? ??? ??? ??? ?   ? ??? ??? ??? H   ??N ??N ??N ???    ???  N
??N ???????    ??N ??N ?   ??N    ??N ??? ??N ??? H   ??N   N     ??N        ??N     ??N  ? 
   ???         ??N ???                  ??? ?? ??? ??? ??N     ??? ??? ??? ??? ??????N 
???   ? ??? ??? ??? ??? ???                 ????   ???                     ??? ???     ???
???                 ???     ??N ???   ? ??? ??? ???   ? ???   ? ??? ???   ?   = ??????? ??? ??????N
   ???   N ??????? ??????? ??? ???????    ??? ??N        ???    ???    ??N ??? ???     ??N 
??? ???????  N       N   ?                 ??N ???    ??N         ??? ??N ??N     ??? @ ??   ???
??? ??? ??? ??? ???             ???     ?  ???    ??? ??? ???     ??? ??? ???     ??? ??? ??? ???
??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ???????
??? ??? ??? ??? ??? ??? ???  ?? ???                  E   G       ???        ???     F   G    ??
??N        ??? ??? ?   ?   ???   N Z   _   ??? ??? U   ^   ???     ;   ?   ??? ??? ?   ?   ??? ???
<   ?   ??? ??N G   G   ???    F   G   ??? ??? E   G   ??? ??? 5   E   ?   U ????N ??? ??? ???
-   /   ???    <   ?   ??? ??? $   '   ??? ??? F   G   ???    E   G                           ???
??? ??? ??? ??? ??? ??N ??? ??? ??? ???    ??N ??? ??? ??N ???    ???
???     "   '   ???     -   /   ???     B   G   ???     $   '   ??? ??? ??? ???     <   ?   ??? ???
??? ???     /   E                   ??? ??? ??? ??? ??N ??? ??? ??N     ??N  ?? ??? ???? ??
????       ???           ???                     ??? ??N ??N ??? ??N ??? ??N ?   ??N ??N  N
??? ??        ??? ??? ??? ??? ??? ??? ??N ??? ??N ??N         ??????? " ? c   g    ?? 3
.1.3            ???  ?? :   ?    ?? e?          ?? e??        2.1.3             e?          ??
???        ??? ???                     ??? ???       ??? ?         ???
   ??? ??? !   '   ??? ???        ?  ???     ??? ??? ???       ?  ???     ??? ??? ???      ??
N ???                   ??? e?          .2.0            ???     g   v   ??? h   c   v   A    ?
???     ??? ??? ???       ?  ???     ??? ??? ???           ?N `??               2.2.0   
??   ??  ?N                                        ???    ???      ???           2.2.0 N 
  ???     O   O   ??? ??N K   O   a
                                ?? ??? ??? ??? ??? ??? ??? o  d   ??                  ??? ???
??N       ??? O         ??? ??N ???    ???    ???    ???    ??? ??N ??? ??? ??? ??? ??? ???
??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ??? ???     
??? ??? "   '   ??? ???       ??? G?? "   '   ??? ??? ????       ??N    ??N ??N ??? ??N ??? ???
??? ???  ??  ?????  ??            ??????? ??? ??? ???            ??N ??N ??N ??? ??N ??? ??N ??N
 N ???  N  N       ?     ???  ? ???  ?  ???    ??? ??? ???         ??????? ??N ??? ??? <   ?
??? ??N ??N ??? ???     ??? ??? ??? ???    ??? ??N ???    ???                    ???        ???
??? "   '   ???  ??       ??? ??? "   '   ??? ???    ??? ??? ??N ??? ??? ??? ??? ??? 
???      ?? ??????? ??? "   c   ???        "   ??? ??N "   .   ??? ??N 2   7   ???        
???            ?? ▀??             ??        ??? ???                 ???         ???
???                 ???                     ???
```

Please note the occurance of the ```2.1.3``` and the ```2.2.0``` strings in the output, which are the two installed dotnet core versions on my machine at the time of running this program. Yep, the hello world program was reading those values from the dotnet CLI's stack.

While it's not easy to abuse an undefined behaviour, a clever attacker could use this bug to get read/write access to any memory address from a safe context.

### It affects even simple programs

In F#, any struct implementating an interface or any struct casted to an object will likely to trigger this bug, and the program will misbehave. In other dotnet languages the tail calls are a bit less frequent, but most of the other languages like C# are definitely affected by this bug.

### No simple workaround

There's no simple workaround for this bug. The relese date for the non-preview 3.0 (which will have the fix that is currently on master) is not even announced yet. All of the versions above 2.1 are affected. Targeting dotnet core 2.0 (instead of 2.1 or 2.2) which did not have this bug yet is not a real solution, as if the executing machine doesn't have the 2.0 SDK installed, it will fail forward for a newer version of the runtime library, and will get affected by this bug. Turning off optimization or tail calls is not viable for applications that have performance requirements. It's very hard to avoid writing code that could trigger this JIT bug, as it can be triggered with very simple and common code.

### It's an easy fix

Brian Sullivan has already fixed this on master, and the backport of this to the 2.1 branch is relatively simple, as you can see in #22377 PR

**Comment by Tomius** [comment]
@briansull @RussKeldorph @jashook @AndyAyersMS

**Comment by AndyAyersMS** [comment]
@Tomius @RussKeldorph we can close this now, right?

**Comment by RussKeldorph** [comment]
Yes, I think so.  #21804 should track it sufficiently.
